CMakeLists.txt

```diff
@@ -204,7 +204,9 @@ endif()
 zephyr_compile_options("SHELL: $<TARGET_PROPERTY:compiler,imacros> ${AUTOCONF_H}")
 
 # @Intent: Set compiler specific flag for bare metal freestanding option
-zephyr_compile_options($<TARGET_PROPERTY:compiler,freestanding>)
+if(NOT CONFIG_SOC_POSIX)
```

Comment:
Could we not find a better way of doing this, than to have very specific tests/cases in the top level CMakeLists.txt file?

Reply:
I completely agree, this is not something i'm fond of, more a short term solution.

We need somehow we can describe if a build is targeting a hosted or freestanding system, and that would probably be the SoC that should set this.

I'm considering something like:
```
config SOC_BUILD_FREESTANDING
        default y
        help
          Prompt less configuration option for SoCs to define 
          if compilation should be build as hosted (setting=n)
          or freestanding application (setting=y).
```

But I haven't taken a deeper look into the most appropriate long term solution, so therefore the current proposal.

Reply:
Yes, something like that, but did we not at some point have different toolchain configurations, depending upon building for target or host? We have configurations named target.cmake and generic.cmake, would it not be possible to utilize that, such as to not modify top level CMakeLists.txt with specific configuration flags?

Reply:
No, the target and generic files are for having just the basic in place for DT.
https://github.com/zephyrproject-rtos/zephyr/commit/a0a63ac3444800de9542fd47e23205ef0ff75468
> Split up the toolchain configuration into two phases, generic and
target. The 'generic' phase configures the toolchain just enough to be
able to preprocess DT files. The 'target' phase completes the
configuration with target-specific configuration.

so this is not related to hosted vs. freestanding application.
Of course we could repurpose those files, but as Zephyr is intended for embedded, i.e. freestanding, then I think it would be too much.

For sure we need to handle building hosted or freestanding in a better way, but especially with llvm, things are not so simple, as it can build both hosted and freestanding using same toolchain. (which gcc also can, but with gcc you have dedicated compiler for cross-compiling --> freestanding).
And on gcc we have the `host` toolchain which simply uses compiler `host-gcc`, but that is not so elegant and I would prefer not to extend that principle into llvm.

I hope you see why I decided a short term proposal that simply uses what is already there, and then look into a more proper solution long term.

Keep posting good suggestions :smile:

Reply:
> No, the target and generic files are for having just the basic in place for DT.

Yes, I see, I had forgotten that. True, that has no relation to host or target selection, then. It might though, be an idea to make host and target selection similar to the DT case, and add a host.cmake file for property settings which are different from target. It could even take the target.cmake as a basis, and then modify the needed properties.
But, as you state:
>  but as Zephyr is intended for embedded, i.e. freestanding

I do agree that it will probably be too much, and actually complicate the process more, which is exactly what we are trying to avoid.

You are correct, it was the `host-gcc` way/split I was thinking about, but that is probably not a good solution either. 

I agree that your current simple solution for llvm and posix can be a start. I do however think, that we should plan to reconsider that at a later state, so that we do not end up with having many special ifs in the top level CMakeLists.txt file, or even many different kconfigs, which will make it difficult/impossible to add new/other toolchains later on.

With regards to:

> config SOC_BUILD_FREESTANDING

I think that is to specific, and it may be better to simply state the intended target, such as host or target with:
> config SOC_TARGET_HOST

or something along those lines. But we can take that at another time, so we do not stall basic llvm for too long.